### PR TITLE
Add support for registering new implementations of KVSender.

### PR DESCRIPTION
**Pull request by petermattis**
Fix support for the "--insecure" flag.

### CODE REVIEW

**Comment by bdarnell** [comment]
LGTM

File: client/sender.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+type newSenderFunc func(u *url.URL, ctx *base.Context) (KVSender, error)
+
+var sendersMu sync.Mutex
+var senders = map[string]newSenderFunc{}
+
+// RegisterSender registers the specified function to be used for
+// creation of a new sender when the specified scheme is encountered.
+func RegisterSender(scheme string, f newSenderFunc) {
+	sendersMu.Lock()
+	sendersMu.Unlock()
```
Comment (tamird) [inline client/sender.go]: defer?

**Comment by tamird** [comment]
LGTM

File: client/sender.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	if _, ok := senders[scheme]; ok {
+		log.Fatalf("sender already registered for \"%s\"", scheme)
+	}
+	senders[scheme] = f
+}
+
+func newSender(u *url.URL, ctx *base.Context) (KVSender, error) {
+	sendersMu.Lock()
+	defer sendersMu.Unlock()
+	f := senders[u.Scheme]
```
Comment (tamird) [inline client/sender.go]: pedantic: f == nil is not the same as the key being absent, and it may be worth distinguishing between the two

**Comment by BramGruneir** [comment]
This is PR #1000!  Awesome!

File: client/db.go [hunk 3]
```diff
@@ -140,11 +139,16 @@ type DB struct {
...
 //
-// The rpc and rpcs schemes use a variant of Go's builtin rpc library for
+// The URL scheme (<sender>) specifies which transport to use for talking to
+// the cockroach cluster. Currently allowable values are: http, https, rpc,
+// rpcs. The rpc and rpcs senders use a variant of Go's builtin rpc library for
 // communication with the cluster. This protocol is lower overhead and more
-// efficient than http.
+// efficient than http. The decision between the encrypted (https, rpcs) and
+// unencrypted senders (http, rpc) depends on the settings of the cluster. A
+// given cluster will not support both encrypted and unencrypted traffic
```
Comment (spencerkimball) [inline client/db.go]: s/will not support both encrypted and unencrypted traffic simultaneously/supports either encrypted or unencrypted traffic, but not both/

File: server/cli/kv.go [hunk 4]
```diff
@@ -40,7 +40,8 @@ var osStderr = os.Stderr
 func makeDBClient() *client.DB {
 	// TODO(pmattis): Initialize the user to something more
 	// reasonable. Perhaps Context.Addr should be considered a URL.
-	db, err := client.Open("https://root@" + util.EnsureHost(Context.Addr) +
+	db, err := client.Open(Context.RequestScheme() +
+		"://root@" + util.EnsureHost(Context.Addr) +
 		"?certs=" + Context.Certs)
```
Comment (spencerkimball) [inline server/cli/kv.go]: Will the "?certs=X" be ignored if insecure is true in the context?

Comment (spencerkimball) [inline server/cli/kv.go]: Maybe we need a unittest for insecure using the kv command line tool.

**Comment by spencerkimball** [comment]
LGTM

File: client/sender.go (hunk 1 above)
Comment (petermattis) [inline client/sender.go]: Oh my! Good catch.

File: client/sender.go (hunk 2 above)
Comment (petermattis) [inline client/sender.go]: Yep, understood. In this context I don't see why I need to distinguish between the two (yet).

File: client/db.go (hunk 3 above)
Comment (petermattis) [inline client/db.go]: Done.

File: client/sender.go (hunk 2 above)
Comment (tamird) [inline client/sender.go]: it strikes me as a bit weird that this differs from the check in `RegisterSender` which only looks for the presence of the key - if i set a key to nil i can never unset it =/

File: client/sender.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	if f == nil {
+		return nil, fmt.Errorf("no sender registered for \"%s\"", u.Scheme)
```
Comment (tamird) [inline client/sender.go]: nit: should this `log.Fatal` like the above? not sure it's worth returning an error in this case

Comment (petermattis) [inline client/sender.go]: I think the asymmetry is ok. `RegisterSender` is called from `init()` functions where it makes sense to `log.Fatal`. `newSender` is called from `client.Open` which already provides for an `error` return value.

File: client/sender.go (hunk 2 above)
Comment (petermattis) [inline client/sender.go]: We can panic above. This is not going to happen in practice.

File: server/cli/kv.go (hunk 4 above)
Comment (petermattis) [inline server/cli/kv.go]: Added a test of the `--insecure` flag which revealed an oddity in our command line flags. Now the "kv" parent command defines the `--addr`, `--certs` and `--insecure` flags which its sub-commands inherit.

Yes, the `"?certs=X"` portion of the URL will be ignored if insecure is true.

**Pull request merged**